### PR TITLE
fix(api) treat bigInt as string all across the system

### LINKED ISSUE #20291: GraphqQL BigInt handling
### Describe the Bug

After further looking into Azri's comment here https://github.com/directus/directus/pull/20263#issuecomment-1787147823 there seem to be a variety of issues when working with BigInt types over GraphQL.

1. filtering bigInt values results in 500 internal server error
```gql
query Test3 {
    test3(filter: { big_int: { _lt: "-9223372036854775800" } }) {
        id
        name
        big_int
    }
}
```

2. filtering NaN values doesnt get caught until the SQL query actually has NaN in it. This happens  for any non-numeric value including an empty string. Same behavior occurs for creating items and likely updating too then.
```gql
query Test3 {
    test3(filter: { big_int: { _lt: "test+123" } }) {
        id
        name
        big_int
    }
}
```

3. creating items with valid BigInt numbers fails with a validation error
```gql
mutation Create_test3_item {
    create_test3_item(data: { big_int: "-9223372036854775800", name: "test" }) {
        name
        big_int
    }
}
```

All these likely have the same root cause, which i suspect is the BigInt implementation here https://github.com/directus/directus/blob/main/api/src/services/graphql/types/bigint.ts which attempts to cast BigInt values to a Number defeating the purpose of using the BigInt type.

### To Reproduce

1. create a collection and add a BigInt field
2a. try to create new item using GraphQL (with a big int value outside the range of a regular Number)
2b. try to filter against a BigInt value (with the value outside the range of a regular Number)
2c. try filtering or creating an item using a string that is not a number

### Directus Version

10.7.1

### Hosting Strategy

Self-Hosted (Docker Image)

### PR DESCRIPTION
**Pull request by mahendraHegde**
## Scope

What's changed:

- remove explicit handling of bigInt on api side, treat it as string and let DB do the casting

## Potential Risks / Drawbacks
- tested with PG and sqlite but not sure if any DB expect the casting as part of query binding itself
- knex doesnt[ parse bigInt](https://github.com/knex/knex/issues/3553), so the queries are sent as `xxxxn` to DB, hence we dont have an option but to use string

## Review Notes / Questions

- found  directus/docs#311 while working on this, as it requires some discussion i created a separate issue
- UI inputs are not handling bigInt as well, its reading the saved data well, but pasting a bigint will result in loosing the precision because of implicit conversion to number(caused by input type=number)

---

Fixes directus/directus#20291

#### create 
Screenshot 2023-12-26 at 9 23 42 PM

#### filter
Screenshot 2023-12-26 at 10 36 07 PM

### invalid
Screenshot 2023-12-26 at 10 36 50 PM

### CODE REVIEW

**Comment by mahendraHegde** [comment]
@br41nslug  ready to be reviewed

**Comment by br41nslug** [comment]
> UI inputs are not handling bigInt as well, its reading the saved data well, but pasting a bigint will result in loosing the precision because of implicit conversion to number(caused by input type=number)

This is being addressed in https://github.com/directus/directus/issues/20287

**Review by br41nslug — COMMENTED** [review]
Thanks for the great PR! Works like a charm 😄 

Would only love to see that tiny constant thing addressed so we dont need a big block of comments explaining why these are hardcoded.

File: api/src/services/graphql/types/bigint.ts [hunk 1]
```diff
@@ -30,13 +30,29 @@ export const GraphQLBigInt = new GraphQLScalarType({
...
 	}
 
-	return value;
+	if (!Number.isSafeInteger(intValue)) {
+		// If the input is not a safe integer, its a big int, so return it as string,
+		// because currently string is the best way to handle big int due to knex limitations and JSON.stringify not able to serialise bigInt
+
+		//its needed to use hardcoded string instead of  2**63-1 directly as it will first convert to number hence loose the precise value
+		const minVal = BigInt('-9223372036854775808'); //-1*(2**63)
+		const maxVal = BigInt('9223372036854775807'); //(2**63-1)
```
[the remark this answers is not in the document]
Comment (mahendraHegde) [reply api/src/services/graphql/types/bigint.ts]: thanks for the great suggestion, dint really think about it. updated the PR

File: .changeset/small-poems-deny.md [hunk 2]
```diff
@@ -0,0 +1,5 @@
+---
+'@directus/api': minor
+---
+
+Fixed issue of not being able to insert or filter bigInt fields in graphQL api
```
Comment (paescuj) [inline .changeset/small-poems-deny.md:5]:
```suggestion
Fixed issue of not being able to insert or filter big integer fields via GraphQL
```

**Review by br41nslug — APPROVED** [review]

**Pull request merged**